Start on the partitioning strategy docs (#179)
IEvangelist/azure-cosmos-dotnet-repository: docs/partitioning-setup.md

# Partitioning Setup

Partitioning is a huge part of Cosmos DB. There are many different strategies you may want to take. Such as, splitting different types of item's across containers, sharing some `IItem`'s in one container and not others and finally, in some cases, even putting all `IItem` types in a single container.

## Default Strategy

Adding the Azure Cosmos Repository with no real configuration, like below. Will result in a simple strategy being assumed on your behalf. This will leave the option's configured in there default state. This means a property called `ContainerPerItemType` is set to false. This means that the Azure Cosmos Repository will store _all_ `IItem`'s in the same container, setting all the `IItem`'s within to use the partition key `/id`.

> ```c#

Reviewer: ```suggestion
```csharp
```

#### Customer Order System Example

In the below example we have two `IItem` types defined. One is the `Customer` item which is configured to be stored in the `customers` container, partitioned by `/emailAddress`. The second is the `Order` item, partitioned by `/customerId`.

```csharp
using Microsoft.Azure.CosmosRepository;

var builder = WebApplication.CreateBuilder(args);

builder.Services.AddCosmosRepository(options => 
{
    options.ContainerPerItemType = true;

    options.ContainerBuilder.Configure<Customer>(userContainerOptions => 
    {    
        userContainerOptions.WithContainer("customer")
        userContainerOptions.WithPartitionKey("/emailAddress")
    });

    options.ContainerBuilder.Configure<Order>(orderContainerOptions => 
    {    
        orderContainerOptions.WithContainer("orders")
        orderContainerOptions.WithPartitionKey("/customerId")
    });
});

var app = builder.Build();

app.MapGet("/", () => 
    "Sample e-commerce Partitioning Strategy");

app.Run();
```

#### Inventory Example

> The below example shows how two or more `IItem` types can share a container and a partitioning strategy. This define a `Stock` item which might hold the name, price, size and maybe weight of a piece of `Stock`. This is partitioned by the `/stockReferenceNumber`. The `StockRecord` item is a collection of `StockRecord`'s, this might be when a piece of stock was added or removed from the inventory, when this occurred and maybe a reason why. This is stored in the _same_ container as the `Stock`'s history that is recording. It also shares the same partitioning strategy. In the example below it means that all `Stock` and it's `StockRecord`'s are stored in the same logical partition in Cosmos DB.

Author: ```suggestion
The below example shows how two or more `IItem` types can share a container and a partitioning strategy. This defines a `Stock` item that might hold the name, price, size, and maybe weight. This is partitioned by the `/stockReferenceNumber`. The `StockRecord` item is a collection of `StockRecord`'s. This record represents when a piece of stock was added or removed from the inventory, when this occurred, and maybe a reason why. This is stored in the _same_ container as the `Stock` item whose history it is recording. It also shares the same partitioning strategy. In the example below it means that all `Stock` and its `StockRecord`'s are stored in the same logical partition in Cosmos DB.
```
